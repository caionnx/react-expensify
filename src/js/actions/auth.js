import { firebase, googleAuthProvider, createUser } from '../firebase/firebase'

export const login = (uid) => ({
  type: 'LOGIN',
  uid
})

export const startCreateUser = (email, password) => () => {
  return createUser(email, password)
}

export const startGoogleLogin = () => () => {
  return firebase.auth().signInWithPopup(googleAuthProvider)
}

export const logout = () => ({
  type: 'LOGOUT'
})

export const startLogout = () => () => {
  return firebase.auth().signOut()
}
