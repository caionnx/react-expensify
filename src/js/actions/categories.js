import database from '../firebase/firebase'

// REMOVE_CATEGORY
export const removeCategory = (id) => ({
  type: 'REMOVE_CATEGORY',
  id
})

export const startRemoveCategory = ({ id } = {}) => (dispatch, getState) => {
  const uid = getState().auth.uid

  return database.ref(`users/${uid}/categories/${id}`).remove().then(() => {
    dispatch(removeCategory(id))
  })
}

// ADD_CATEGORY
export const addCategory = (category) => ({
  type: 'ADD_CATEGORY',
  category
})

export const startAddCategory = (category = {}) => (dispatch, getState) => {
  const uid = getState().auth.uid
  const { id, value } = category

  return database.ref(`users/${uid}/categories/${id}`).set(value).then(() => {
    dispatch(addCategory(category))
  })
}

export const startRegisterCategories = (categories, uid) => {
  return database.ref(`users/${uid}/categories`).set({...categories})
}

// SET_CATEGORIES
export const setCategories = (categories) => ({
  type: 'SET_CATEGORIES',
  categories
})

export const startSetCategories = () => (dispatch, getState) => {
  const uid = getState().auth.uid
  const transformToArray = (object) => {
    const arr = []
    object.forEach(el => {
      arr.push({
        id: el.key,
        value: el.val()
      })
    })

    return arr
  }

  return database.ref(`users/${uid}/categories`).once('value').then((snapshot) => {
    let categories
    if (snapshot.hasChildren()) {
      categories = transformToArray(snapshot)
      dispatch(setCategories(categories))
    } else {
      database.ref(`categories`).once('value').then((defaultSnapshot) => {
        categories = transformToArray(defaultSnapshot)
        dispatch(setCategories(categories))

        return defaultSnapshot.hasChildren() &&
          startRegisterCategories(defaultSnapshot.val(), uid)
      })
    }
  })
}
