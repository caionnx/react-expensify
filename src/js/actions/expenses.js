import database from '../firebase/firebase'
import modeOperator from './loginModeOperator'

// ADD_EXPENSE
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
})

export const startAddExpense = (expenseData = {}) => (dispatch, getState) => {
  const auth = getState().auth
  const { uid } = auth
  const {
    description = '',
    note = '',
    amount = 0,
    createdAt = 0,
    category = 'none'
  } = expenseData

  const expense = { description, note, amount, createdAt, category }

  return modeOperator(auth, {
    anonymous: () => dispatch(addExpense({
      id: String(Math.floor(Math.random() * 100)),
      ...expense
    })),
    connected: () => database.ref(`users/${uid}/expenses`).push(expense).then(ref => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }))
    })
  })
}

// REMOVE_EXPENSE
export const removeExpense = (id) => ({
  type: 'REMOVE_EXPENSE',
  id
})

export const startRemoveExpense = ({ id } = {}) => (dispatch, getState) => {
  const auth = getState().auth
  const { uid } = auth

  return modeOperator(auth, {
    anonymous: () => dispatch(removeExpense(id)),
    connected: () => database.ref(`users/${uid}/expenses/${id}`).remove().then(() => {
      dispatch(removeExpense(id))
    })
  })
}

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
})

export const startEditExpense = (id, updates) => (dispatch, getState) => {
  const auth = getState().auth
  const { uid } = auth

  return modeOperator(auth, {
    anonymous: () => dispatch(editExpense(id, updates)),
    connected: () => database.ref(`users/${uid}/expenses/${id}`).update(updates).then(() => {
      dispatch(editExpense(id, updates))
    })
  })
}

// SET_EXPENSES
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
})

export const startSetExpenses = () => (dispatch, getState) => {
  const auth = getState().auth
  const { uid } = auth

  return modeOperator(auth, {
    anonymous: () => dispatch(setExpenses([])),
    connected: () => database.ref(`users/${uid}/expenses`).once('value').then((snapshot) => {
      const expenses = []
      snapshot.forEach(element => {
        expenses.push({
          id: element.key,
          ...element.val()
        })
      })

      dispatch(setExpenses(expenses))
    })
  })
}
