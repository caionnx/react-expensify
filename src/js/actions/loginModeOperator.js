export default (auth, {anonymous = () => null, connected = () => null} = {}) => {
  const { isAnonymous } = auth
  return isAnonymous
    ? Promise.resolve(anonymous())
    : connected()
}
