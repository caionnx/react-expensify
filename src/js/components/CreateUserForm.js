import React from 'react'
import { connect } from 'react-redux'
import { startCreateUser } from '../actions/auth'
import PropTypes from 'prop-types'

export class CreateUserForm extends React.Component {
  state = {
    error: false,
    loading: false
  }

  isEmail = (email) =>
    /(.+)@(.+){2,}\.(.+){2,}/.test(email)

  isPassword = (password) =>
    /^(?=.*\d)(?=.*[a-z])(?=.*[A-Z]).{6,8}$/.test(password)

  onFormSubmit = (ev) => {
    ev.preventDefault()
    const formData = new FormData(ev.target) // eslint-disable-line
    const email = formData.get('email')
    const confirmEmail = formData.get('confirm-email')
    const password = formData.get('password')

    if (!this.isEmail(email)) {
      this.setState(() => ({ error: 'Invalid email address.' }))
      return
    }

    if (email !== confirmEmail) {
      this.setState(() => ({ error: 'Please, confirm your email.' }))
      return
    }

    if (!this.isPassword(password)) {
      this.setState(() => ({ error: 'Password must be min 6 characters and max 8 characters. Must include one upper case letter, one lower case letter, and one numeric digit.' }))
      return
    }

    this.setState(() => ({ loading: true }))
    this.props.startCreateUser(email, password).catch((fail) => {
      const error = fail.code === 'auth/email-already-in-use'
        ? fail.message
        : 'Fail to create account.'

      this.setState(() => ({ error, loading: false }))
    })
  }

  render () {
    const { error, loading } = this.state
    const { goBackFunction } = this.props

    return (
      <div>
        <p>Create your account.</p>
        <form className='form' onSubmit={this.onFormSubmit}>
          <input placeholder='Email' aria-label='Email' type='text' name='email' required className='text-input' />
          <input placeholder='Confirm Email' aria-label='Confirm Email' type='text' name='confirm-email' required className='text-input' />
          <input placeholder='Password' aria-label='Password' type='password' name='password' required className='text-input' />
          { error && <p className='form__error'>{error}</p> }
          <button type='submit' className='button' disabled={loading}>
            { loading ? 'Submiting..' : 'Create' }
          </button>
          { goBackFunction &&
            <a href='#' className='button button--secondary' onClick={goBackFunction}>Cancel</a> }
        </form>
      </div>
    )
  }
}

CreateUserForm.propTypes = {
  startCreateUser: PropTypes.func,
  goBackFunction: PropTypes.func
}

const mapDispatchToProps = (dispatch) => ({
  startCreateUser: (emai, password) => dispatch(startCreateUser(emai, password))
})

export default connect(undefined, mapDispatchToProps)(CreateUserForm)
