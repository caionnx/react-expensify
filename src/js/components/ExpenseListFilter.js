import React from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import { setTextFilter, sortByDate, sortByAmount } from '../actions/filters'

const onChangeSortBy = (dispatch, e) =>
  e.target.value === 'date'
    ? dispatch(sortByDate(e.target.value))
    : dispatch(sortByAmount(e.target.value))

export const ExpenseListFilter = (props) => (
  <div>
    <input
      type='text'
      value={props.filters.text}
      onChange={(e) => props.dispatch(setTextFilter(e.target.value))} />
    <select
      onChange={(e) => onChangeSortBy(props.dispatch, e)}
      value={props.filters.sortBy}>
      <option value='date'>Date</option>
      <option value='amount'>Amount</option>
    </select>
  </div>
)

ExpenseListFilter.propTypes = {
  dispatch: PropTypes.func,
  filters: PropTypes.shape({
    sortBy: PropTypes.string,
    text: PropTypes.string
  })
}

const mapStateToProps = (state) => ({
  filters: state.filters
})

export default connect(mapStateToProps)(ExpenseListFilter)
