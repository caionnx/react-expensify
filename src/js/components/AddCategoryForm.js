import React from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import slugify from 'slugify'
import { startAddCategory } from '../actions/categories'

export class AddCategoryForm extends React.Component {
  state = {
    error: ''
  }

  onAddCategory = (event) => {
    event.preventDefault()
    const node = event.target.querySelector(`#new-category`)
    const value = node && node.value
    const id = slugify(value, { replacement: '-', remove: null, lower: true })

    if (id === '') {
      this.setState(() => ({ error: 'Please provide category name.' }))
      return
    }
    if (this.props.categories.filter((item) => item.id === id).length) {
      this.setState(() => ({ error: 'Category already registered.' }))
      return
    }

    this.setState(() => ({ error: '' }))
    this.props.startAddCategory({ id, value })
    node.value = ''
  }

  render () {
    const { error } = this.state

    return (
      <form onSubmit={this.onAddCategory}>
        {error && <p className='form__error'>{error}</p>}
        <div className='input-group'>
          <div className='input-group__item'>
            <input id='new-category' className='text-input' placeholder='New Category' type='text' />
          </div>
          <div className='input-group__item'>
            <button className='button'>Add Category</button>
          </div>
        </div>
      </form>
    )
  }
}

AddCategoryForm.propTypes = {
  categories: PropTypes.array,
  startAddCategory: PropTypes.func.isRequired
}

const mapStateToProps = ({ categories }) => ({
  categories
})

const mapDispatchToProps = (dispatch) => ({
  startAddCategory: (category) => dispatch(startAddCategory(category))
})

export default connect(mapStateToProps, mapDispatchToProps)(AddCategoryForm)
