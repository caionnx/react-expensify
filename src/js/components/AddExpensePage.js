import React from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import ExpenseForm from './ExpenseForm'
import { startAddExpense } from '../actions/expenses'

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startAddExpense(expense).then(() => {
      this.props.history.push('/')
    })
  }
  render () {
    return (
      <div>
        <div className='page-header'>
          <div className='content-container'>
            <h1 className='page-header__title'>Add Expense</h1>
          </div>
        </div>
        <div className='content-container'>
          <ExpenseForm
            onSubmit={this.onSubmit}
          />
        </div>
      </div>
    )
  }
}

AddExpensePage.propTypes = {
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }),
  startAddExpense: PropTypes.func.isRequired
}

const mapDispatchToProps = (dispatch) => ({
  startAddExpense: (expense) => dispatch(startAddExpense(expense))
})

export default connect(undefined, mapDispatchToProps)(AddExpensePage)
