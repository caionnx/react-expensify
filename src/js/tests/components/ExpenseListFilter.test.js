import React from 'react'
import { shallow } from 'enzyme'
import { ExpenseListFilter } from '../../components/ExpenseListFilter'
import { filters } from '../fixtures/filters'
import { setTextFilter, sortByDate, sortByAmount } from '../../actions/filters'

let wrapper, dispatch

beforeAll(() => {
  dispatch = jest.fn()
  wrapper = shallow(
    <ExpenseListFilter
      dispatch={dispatch}
      filters={filters}
    />
  )
})

test('should render ExpenseListFilter correctly', () => {
  expect(wrapper).toMatchSnapshot()
})

test('should dispatch text filter action', () => {
  const valueOfEvent = 'test'
  const action = setTextFilter(valueOfEvent)
  wrapper.find('input').simulate('change', {target: { value: valueOfEvent }})

  expect(dispatch).toHaveBeenLastCalledWith(action)
})

test('should dispatch sort by date action', () => {
  const valueOfEvent = 'date'
  const action = sortByDate(valueOfEvent)
  wrapper.find('select').simulate('change', {target: { value: valueOfEvent }})

  expect(dispatch).toHaveBeenLastCalledWith(action)
})

test('should dispatch sort by amount action', () => {
  const valueOfEvent = 'amount'
  const action = sortByAmount(valueOfEvent)
  wrapper.find('select').simulate('change', {target: { value: valueOfEvent }})

  expect(dispatch).toHaveBeenLastCalledWith(action)
})
