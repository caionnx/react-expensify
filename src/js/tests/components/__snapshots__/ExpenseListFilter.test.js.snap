// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListFilter correctly 1`] = `
<div>
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
  <select
    onChange={[Function]}
    value="date"
  >
    <option
      value="date"
    >
      Date
    </option>
    <option
      value="amount"
    >
      Amount
    </option>
  </select>
</div>
`;
